File: 2392-successful-pairs-of-spells-and-potions/2392-successful-pairs-of-spells-and-potions.cpp
class Solution {
public:
    // int lowerBound( int l ,  int r , vector<int>& potions, long long minpotion){
    //     int ps = -1;
    //     while( l <= r){
    //         int mid = l  + ( r-l ) / 2;
    //         if(potions[mid] >= minpotion){
    //             ps = mid;
    //             r = mid -1;

    //         }else{ l = mid;}
    //     }
    //     return ps;
    // }

    vector<int> successfulPairs(vector<int>& spells, vector<int>& potions, long long success) {
        // int m = spells.size(); 
        // int n = potions.size();
        // sort(begin(potions) , end(potions));

        // vector<int> answers;

        // for(int i = 0 ; i <  m ; i++){
        //     int spell = spells[i];
        //     long long minpotion = ceil((1.0 * success) / spell);

        //     int index  = lowerBound(0 , n - 1 , potions , minpotion);
        //     int count = n - index;

        //     answers.push_back(count);
        // }
        // return answers;
        int m = spells.size(); 
        int n = potions.size();
        sort(begin(potions) , end(potions));

        vector<int> answers;

        for(int i = 0 ; i <  m ; i++){
            int spell = spells[i];
            long long minpotion = ceil((1.0 * success) / spell);

            int index  = lower_bound(begin(potions) , end(potions), minpotion) - begin(potions);
            int count = n - index;

            answers.push_back(count);
        }
        return answers;
    }
};

File: 2392-successful-pairs-of-spells-and-potions/2392-successful-pairs-of-spells-and-potions.java
class Solution {
    public int[] successfulPairs(int[] spells, int[] potions, long success) {
        // Arrays.sort(potions);
        // int n = spells.length, m = potions.length;
        // int[] result = new int[n];

        // for (int i = 0; i < n; i++) {
        //     long minPotion = (success + spells[i] - 1) / spells[i]; // Avoid overflow
        //     int index = binarySearch(potions, minPotion);
        //     result[i] = m - index;
        // }

        // return result;
        var l = potions.length;
        Arrays.sort(potions);
        var results = new int[spells.length];

        for (int i=0; i<spells.length; i++) { 
            var s = spells[i];
            int start = 0, end = l, magicIndex = -1;
            while (start < end) {
                var mid = start + (end - start) / 2;
                long test = (long) s * potions[mid];
                if (test >= success) {
                    magicIndex = mid;
                    end = mid;
                } else {
                    start = mid + 1;
                }
            }
            if (magicIndex != -1) {
                results[i] = l - magicIndex;
            } else {
                results[i] = 0;
            }
        }
        return results;
    }
    // private int binarySearch(int[] potions, long minPotion) {
    //     int low = 0, high = potions.length - 1;
    //     while (low <= high) {
    //         int mid = low + (high - low) / 2;
    //         if (potions[mid] < minPotion) {
    //             low = mid + 1;
    //         } else {
    //             high = mid - 1;
    //         }
    //     }
    //     return low;
    // }


}
